src/core/meshcop/dataset_manager.cpp

```diff
@@ -279,7 +278,7 @@ void DatasetManager::SendSet(void)
     Dataset          dataset(GetType());
 
     VerifyOrExit(!mCoapPending, error = kErrorBusy);
-    VerifyOrExit(Get<Mle::MleRouter>().IsAttached(), error = kErrorInvalidState);
+    VerifyOrExit(Get<Mle::MleRouter>().IsChild() || Get<Mle::MleRouter>().IsRouter(), error = kErrorInvalidState);
```

Comment:
@jwhui  Will this reject if the leader sends MGMT_SET.req to itself?

Reply:
Yeah, the idea is that the Leader is authoritative and calls to `SendSet()` should be ignored on the Leader. In what case do we need to call `SendSet()` on the Leader to update itself?

Reply:
Yeah, but it should be legal, right? It sounds more reasonable that the leader should be able to send requests to itself.

Thoughts?
@librasungirl @abtink

Reply:
Possible use case - when the administrator want to commission the network to migrate to another network, there might be chance that the device that receives such command is a Leader.

Currently ot-br-posix will always sending pending dataset regardless of the device is leader in [thread_helper](https://github.com/openthread/ot-br-posix/blob/main/src/utils/thread_helper.cpp#L500)

```
void ThreadHelper::AttachAllNodesTo(const std::vector<uint8_t> &aDatasetTlvs, ResultHandler aHandler)
{
....
    SuccessOrExit(error = otDatasetSendMgmtPendingSet(mInstance, &emptyDataset, datasetTlvs.mTlvs, datasetTlvs.mLength,
                                                      MgmtSetResponseHandler, this));
...
}
```

To solve it, either allowing leader sends mgmt to itself, or may updating `thread_helper` to set pending dataset directly on leader local.

//cc// @jinran-google @bukepo for vis.

Reply:
@wgtdkp @librasungirl , how does calling `otDatasetSendMgmtPendingSet()` lead to calling `SendSet()`?

Reply:
@wgtdkp 's on device test log `Failed to send Dataset set to leader: InvalidState` seems it would somehow enter `SentSet()`,  I didn't carefully look into the calling path, and yes, you are right, it's `SendSetRequest()` that is called in `otDatasetSendMgmtPendingSet()`. 

 Will follow up with @wgtdkp on his setup and test procedure to see how the log occurs offline. Temporarily close this discussion now.